lib/lrmd/proxy_common.c

```diff
@@ -111,4 +109,185 @@ remote_proxy_free(gpointer data)
+
+    } else {
+        crm_trace("Passing event back to %.8s on %s: %.200s", proxy->session_id, proxy->node_name, buffer);
+        remote_proxy_relay_event(proxy, xml);
+    }
+    free_xml(xml);
+    return 1;
+}
+
+
+void
+remote_proxy_disconnected(gpointer userdata)
+{
+    remote_proxy_t *proxy = userdata;
 
+    crm_trace("destroying %p", proxy);
+
+    proxy->source = NULL;
+    proxy->ipc = NULL;
+
+    if(proxy->lrm) {
+        remote_proxy_notify_destroy(proxy->lrm, proxy->session_id);
+        proxy->lrm = NULL;
+    }
+
+    g_hash_table_remove(proxy_table, proxy->session_id);
+}
+
+remote_proxy_t *
+remote_proxy_new(lrmd_t *lrmd, struct ipc_client_callbacks proxy_callbacks,
+                 const char *node_name, const char *session_id, const char *channel)
+{
+    remote_proxy_t *proxy = NULL;
+
+    if(channel == NULL) {
+        crm_err("No channel specified to proxy");
+        remote_proxy_notify_destroy(lrmd, session_id);
+        return NULL;
+    }
+
+    proxy = calloc(1, sizeof(remote_proxy_t));
+
+    proxy->node_name = strdup(node_name);
+    proxy->session_id = strdup(session_id);
+
+    proxy->source = mainloop_add_ipc_client(channel, G_PRIORITY_LOW, 0, proxy, &proxy_callbacks);
+    proxy->ipc = mainloop_get_ipc_client(proxy->source);
+    proxy->lrm = lrmd;
+
+    if (proxy->source == NULL) {
+        remote_proxy_free(proxy);
+        remote_proxy_notify_destroy(lrmd, session_id);
+        return NULL;
+    }
+
+    crm_trace("new remote proxy client established to %s on %s, session id %s",
+              channel, node_name, session_id);
+    g_hash_table_insert(proxy_table, proxy->session_id, proxy);
+
+    return proxy;
+}
+
+void
+remote_proxy_cb(lrmd_t *lrmd, const char *node_name, xmlNode *msg)
+{
+    const char *op = crm_element_value(msg, F_LRMD_IPC_OP);
+    const char *session = crm_element_value(msg, F_LRMD_IPC_SESSION);
+    remote_proxy_t *proxy = g_hash_table_lookup(proxy_table, session);
+    int msg_id = 0;
+
+    /* sessions are raw ipc connections to IPC,
+     * all we do is proxy requests/responses exactly
+     * like they are given to us at the ipc level. */
+
+    CRM_CHECK(op != NULL, return);
+    CRM_CHECK(session != NULL, return);
+
+    crm_element_value_int(msg, F_LRMD_IPC_MSG_ID, &msg_id);
+    /* This is msg from remote ipc client going to real ipc server */
+
+    if (safe_str_eq(op, LRMD_IPC_OP_DESTROY)) {
+        remote_proxy_end_session(proxy);
+
+    } else if (safe_str_eq(op, LRMD_IPC_OP_REQUEST)) {
+        int flags = 0;
+        xmlNode *request = get_message_xml(msg, F_LRMD_IPC_MSG);
+        const char *name = crm_element_value(msg, F_LRMD_IPC_CLIENT);
+
+        CRM_CHECK(request != NULL, return);
+
+        if (proxy == NULL) {
+            /* proxy connection no longer exists */
+            remote_proxy_notify_destroy(lrmd, session);
+            return;
+        } else if ((proxy->is_local == FALSE) && (crm_ipc_connected(proxy->ipc) == FALSE)) {
```

Comment:
Since we removed the is_local branch from this function, we can assume it's false here